File: wkbase/src/main/java/com/chat/base/net/upload/FileRequestBody.java
package com.chat.base.net.upload;

import android.os.Handler;
import android.os.Looper;

import androidx.annotation.NonNull;

import com.chat.base.okgo.OkGoUploadOrDownloadProgress;

import java.io.IOException;

import okhttp3.MediaType;
import okhttp3.RequestBody;
import okio.Buffer;
import okio.BufferedSink;
import okio.ForwardingSink;
import okio.Okio;

public class FileRequestBody extends RequestBody {
    private final RequestBody requestBody;
    private long mCurrentLength;
    private final Object tag;
    Handler handler = new Handler(Looper.getMainLooper());

    public FileRequestBody(RequestBody requestBody, Object tag) {
        super();
        this.tag = tag;
        this.requestBody = requestBody;
    }

    @Override
    public long contentLength() throws IOException {
        return requestBody.contentLength();
    }

    @Override
    public MediaType contentType() {
        return requestBody.contentType();
    }

    @Override
    public void writeTo(@NonNull BufferedSink sink) throws IOException {
        final long contentLength = contentLength();
        ForwardingSink forwardingSink = new ForwardingSink(sink) {
            @Override
            public void write(@NonNull Buffer source, long byteCount) throws IOException {
                // 每次写都会来这里
                mCurrentLength += byteCount;
                float f1 = mCurrentLength / (float) contentLength;
                handler.post(() -> OkGoUploadOrDownloadProgress.getInstance().seekProgress(tag, f1));
                super.write(source, byteCount);
            }
        };
        BufferedSink bufferedSink = Okio.buffer(forwardingSink);
        requestBody.writeTo(bufferedSink);
        bufferedSink.flush();
    }

}

File: wkbase/src/main/java/com/chat/base/net/ud/FileRequestBody.kt
package com.chat.base.net.ud

import android.os.Handler
import android.os.Looper
import com.chat.base.utils.WKLogUtils
import okhttp3.MediaType
import okhttp3.RequestBody
import okio.Buffer
import okio.BufferedSink
import okio.ForwardingSink
import okio.Okio
import okio.buffer
import okio.sink
import java.io.IOException


class FileRequestBody(private val requestBody: RequestBody, private val tag: Any) : RequestBody() {
    private var mCurrentLength: Long = 0
    private var bufferedSink: BufferedSink? = null
    var handler = Handler(Looper.getMainLooper())
    override fun contentLength(): Long {
        return requestBody.contentLength()
    }

    override fun isOneShot(): Boolean {
        WKLogUtils.e("是否执行一次")
        return true
    }
    override fun contentType(): MediaType? {
        return requestBody.contentType()
    }

    override fun writeTo(sink: BufferedSink) {
        val contentLength = contentLength()
        WKLogUtils.e("上传总长度$contentLength")
        val forwardingSink: ForwardingSink = object : ForwardingSink(sink) {
            @Throws(IOException::class)
            override fun write(source: Buffer, byteCount: Long) {
                mCurrentLength += byteCount
                val f1 = mCurrentLength / contentLength.toFloat()
                handler.post {
                    var p = (f1 * 100).toInt()
                    WKLogUtils.e("当前总长度$p")
                    if (p > 100) {
                        p = 100
                    }
                    WKProgressManager.instance.seekProgress(tag, p)
                }
                super.write(source, byteCount)
            }
        }
        val bufferedSink: BufferedSink = forwardingSink.buffer()
        requestBody.writeTo(bufferedSink)
        bufferedSink.flush()
    }
}